L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,390 @@
+class TlsChannelCredentialsOptions final : public TlsCredentialsOptions {
+ public:
+  // Sets the decision of whether to do a crypto check on the server certs.
+  // The default is true.
+  void set_verify_server_certs(bool verify_server_certs);
+
+// Builds TLS Credentials given TLS options.
+std::shared_ptr<ChannelCredentials> TlsCredentials(
+    const TlsChannelCredentialsOptions& options);
+}
+
+std::shared_ptr<ServerCredentials> TlsServerCredentials(
+    const grpc::experimental::TlsServerCredentialsOptions& options);
+```
+
+Here is an example of how to use this API:
+```c++
+/* Create option and set basic security primitives. */
+TlsChannelCredentialsOptions options = TlsChannelCredentialsOptions();
+options.set_verify_server_cert(true);
+options.set_min_tls_version(TLS1_2);
+options.set_max_tls_version(TLS1_3);
+
+// The core credentials APIs are still C, so this is still a grpc_channel_credential
+/* Create TLS channel credentials. */
+std::shared_ptr<ChannelCredentials> creds = TlsCredentials(options);
+```
+
+### Credential Reloading
+Credential reloading will be implemented via a `CertificateProvider`. This provider is responsible for sourcing key material and supplying it to the internal stack via `SetKeyMaterials`.
+We provide a `StaticDataCertificateProvider` and a `FileWatcherCertificateProvider` that should cover many use cases.
+If these providers do not support the intricacies for a specific use case, a user can provide their own implementations of the `CertificateProviderInterface`.
+
+```c
+/* Opaque types. */
+// A struct that stores the credential data presented to the peer in handshake
+// to show local identity. The private_key and certificate_chain must be PEM
+// encoded and should always match.
+struct IdentityKeyCertPair {
+  std::string private_key;
+  std::string certificate_chain;
+};
+
+
+/** ------------------------------------ Provider ------------------------------------ */
+/** Interface for a class that handles the process to fetch credential data.
+*/
+class CertificateProviderInterface {
+ public:
+  virtual ~CertificateProviderInterface() = default;
+  // Provider implementations MUST provide a WatchStatusCallback that will be
+  // called by the internal stack. This will be invoked when a new certificate
+  // name is watched by a newly registered watcher, or when a certificate name
+  // is no longer watched by any watchers.  Note that when the callback shows a
+  // cert is no longer being watched, corresponding certificate data will be
+  // deleted from the internal cache, and any corresponding errors will be
+  // cleared, if there are any. This means that if the callback subsequently
+  // says the same cert is now being watched again, the provider must re-provide
+  // the credentials or re-invoke the errors to indicate a successful or failed
+  // reloading.
+  //
+  // For the parameters in the callback function:
+  // cert_name The name of the certificates being watched.
```

Comment:
Same here, what does a "name" mean? Is it a file path?

Reply:
It is user provided, I think the text in the options struct explains this? I can add more, WDYT?